### ISSUE TITLE
clang-tidy fails to build in the latest code

### ISSUE THREAD

**Issue by krasznaa** [intel/llvm#3011]
Hi,

I'm not sure why it would only show up in my build, but I get the following build failure with the current HEAD of the code:

```
[3321/6558] Building CXX object tools/clang/tools/extra/clang-tidy/altera/CMakeFiles/obj.clangTidyAlteraModule.dir/SingleWorkItemBarrierCheck.cpp.o
FAILED: tools/clang/tools/extra/clang-tidy/altera/CMakeFiles/obj.clangTidyAlteraModule.dir/SingleWorkItemBarrierCheck.cpp.o 
/opt/lcg/gcc/8.3.0-eda0e/x86_64-centos7/bin/g++ -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools/clang/tools/extra/clang-tidy/altera -I/root/llvm/clang-tools-extra/clang-tidy/altera -Itools/clang/tools/extra/clang-tidy -I/root/llvm/llvm/../llvm-spirv/include -I/root/llvm/clang/include -Itools/clang/include -Iinclude -I/root/llvm/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-class-memaccess -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -fdiagnostics-color -ffunction-sections -fdata-sections -fno-common -Woverloaded-virtual -fno-strict-aliasing -O3 -DNDEBUG -std=c++14 -MD -MT tools/clang/tools/extra/clang-tidy/altera/CMakeFiles/obj.clangTidyAlteraModule.dir/SingleWorkItemBarrierCheck.cpp.o -MF tools/clang/tools/extra/clang-tidy/altera/CMakeFiles/obj.clangTidyAlteraModule.dir/SingleWorkItemBarrierCheck.cpp.o.d -o tools/clang/tools/extra/clang-tidy/altera/CMakeFiles/obj.clangTidyAlteraModule.dir/SingleWorkItemBarrierCheck.cpp.o -c /root/llvm/clang-tools-extra/clang-tidy/altera/SingleWorkItemBarrierCheck.cpp
/root/llvm/clang-tools-extra/clang-tidy/altera/SingleWorkItemBarrierCheck.cpp: In member function 'virtual void clang::tidy::altera::SingleWorkItemBarrierCheck::check(const clang::ast_matchers::MatchFinder::MatchResult&)':
/root/llvm/clang-tools-extra/clang-tidy/altera/SingleWorkItemBarrierCheck.cpp:60:34: error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
       if (Attribute->getXDim() > 1 || Attribute->getYDim() > 1 ||
                                  ^
/root/llvm/clang-tools-extra/clang-tidy/altera/SingleWorkItemBarrierCheck.cpp:60:62: error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
       if (Attribute->getXDim() > 1 || Attribute->getYDim() > 1 ||
                                                              ^
/root/llvm/clang-tools-extra/clang-tidy/altera/SingleWorkItemBarrierCheck.cpp:61:34: error: ISO C++ forbids comparison between pointer and integer [-fpermissive]
           Attribute->getZDim() > 1)
                                  ^
```

(The build is happening inside of a Docker container, that's why the code is under `/root`...)

As you can see from the output as well, the build is done using GCC 8.3.0, on CentOS 7. But I doubt that the OS makes much of a difference for this.

I see that this code was added in: e69e551e0e5fddffb6479da6a2998457104ba9e6 But I don't know why the CI (or any developer's build for that matter) would not fail on it... :confused:

Cheers,
       Attila

**Comment by bader** [comment]
> But I don't know why the CI (or any developer's build for that matter) would not fail on it... 😕

I assume that CI (or any developer) doesn't build clang-tidy project from `intel/llvm` and this project is not enabled by default.

AFAIK, we do not add any functionality to clang-tidy project, so I assume you can skip building it locally and use pre-built binaries (or build it from `main` branch instead of `sycl`).

**Comment by krasznaa** [comment]
Hi @bader,

I'm not sure myself why, but `clang-tidy` **is** built as part of the SYCL toolchain. I ran into this issue with the provided Python build scripts as well. (Even though I usually don't use them.)

And the issue, as I wrote, very much comes from code added by you guys for FPGAs.

https://github.com/intel/llvm/commit/e69e551e0e5fddffb6479da6a2998457104ba9e6

Cheers,
      Attila

**Comment by AaronBallman** [comment]
This is our issue and not a community issue. We recently made a change that's not upstreamed yet that broke this clang-tidy check: https://github.com/intel/llvm/pull/2906

Note how the community version of the attribute is taking a different argument type than our own:
https://github.com/llvm/llvm-project/blob/master/clang/include/clang/Basic/Attr.td#L2251
https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/Attr.td#L2848

**Comment by bader** [comment]
@krasznaa, the issue is supposed to be addressed by #3016. Let us know if you have any issues with building the toolchain.

**Comment by krasznaa** [comment]
Yepp, I was able to build 42c09e3b4e075f4905a774e0d7540bbfe953a367 successfully in my containerised environment. :thumbsup:
